### PR TITLE
PAT integration samples

### PR DESCRIPTION
**Pull request by tungd**
## What's there and what are left to be done:

- [x] Button on Partner site to start the linking process
- [x] API call from Partner site to PAT, redirect handling
- [x] Tenant to store the PAT Link details, and respond with Consent Capture Page url
- [x] Rendering of Consent Capture Page on Tenant site
- [x] Display link constraints on Consent Capture Page, currently there's only Intent and Description
- [x] Approve / Decline (for PAT links, not PAT transactions) button on Consent Capture Page on Tenant site
- [x] Redirect back to Gateway after Approve / Decline (for PAT links, not PAT transactions)
- [ ] Update Nginx config to expose the sample
- [x] (Optional) Accept different PAT Link details (amount, currency, constraints) from a input form, similar to Transaction for testing purposes
- [x] (Optional) Polish UI

### Start the Partner sample

Verify that `pat-link-sample/.env` is filled with correct details, for example, following is my current setup, using DEV authentication server and credentials, but with services running locally:

```
OF_AUTH_ENDPOINT=https://auth.dev.openfabric.co
OF_API_ENDPOINT=http://127.0.0.1:2023/
MERCHANT_CLIENT_ID=4ec3ah44okuu2ita53fq9qei3m
MERCHANT_CLIENT_SECRET=1qbm5oqqj1jbnooch35s7lr12et420ejjfp7vq530sh396kbbqt2
```

```sh
cd pat-link-sample
yarn install
yarn run build
yarn dev
```

### Start the Tenant sample

```sh
cd account-integration-sample
yarn install
yarn start
```

### Local dev proxy

Since the PAT service is not deployed yet, it is necessary to run additional local server to rewrite the request path (`/v1/preapproved_transaction_links` --> `/public/pat/v1/preapproved_transaction_links`) and provide `of-authorization` context for the local server. It is important that this dev proxy is not included within this repository, since this repository is public and we don't want to reveal/expose our system's internal. 

### Ports

Partner sample:3000 --> Dev proxy:2023 --> PAT service: 5001 --> Tenant sample:4000

### CODE REVIEW

File: account-integration-sample/controllers/pat.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        "Content-Type": "application/json",
+      },
+    }
+  );
+
+
+  //
+  res.send(
+    {
+      url: `${of_pat_url}/v1/preapproved_transaction_links/redirect?link_id=${result.data.id}&status=${status}`
```
Comment (vinh0604) [inline account-integration-sample/controllers/pat.js:70]: Tenant doesn't need to configure the redirect URL like this, they just need to use the OF gateway redirect URL submitted earlier as is.

File: account-integration-sample/controllers/pat.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  let reason;
+  if(status == "approved") {
+    reason = "Card is valid for purchasing";
+  } else {
+    reason = "Card is blocked"
+  }
+  const result = await axios.patch(
+    new URL("/v1/preapproved_transaction_links", of_pat_url).toString(),
+    {
+      id: req.body.of_link_ref,
```
Comment (vinh0604) [inline account-integration-sample/controllers/pat.js:53]: `id` can be removed, as only `tenant_link_ref` is required for approve/decline PAT link.

File: account-integration-sample/controllers/pat.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    partner_customer_id: req.body.partner_customer_id,
+    intent: req.body.intent,
+    description: req.body.description,
+    constraints: req.body.constraints,
+    of_link_ref: req.body.of_link_ref
+  })
+  res.send({
+    tenant_link_ref: ref,
+    tenant_customer_id: customerId,
+    consent_capture_page_url: `http://localhost:4000/pat/consent/${ref}`
```
Comment (vinh0604) [inline account-integration-sample/controllers/pat.js:25]: `http://localhost:4000` should be parameterized with the actual domain that going to deploy the Tenant samples,

File: pat-link-sample/src/pages/index.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        description: 'Weekly subscription for goods'
+    })
+    const Router = useRouter();
+    const startLink = (e: MouseEvent<HTMLButtonElement>) => {
+        e.preventDefault()
+        console.log('start linking')
+    }
+
+    const onPayClick = async () => {
+        const transaction = await (
```
Comment (vinh0604) [inline pat-link-sample/src/pages/index.tsx:23]: Please rename to `consentRedirectUrl` to better describe the variable value.

File: pat-link-sample/src/pages/index.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                },
+                body: JSON.stringify({
+                    amount: state.amount,
+                    currency: state.currency,
+                    description: state.description,
+                    frequency: state.frequency
+                }),
+            })
+        ).text();
+        Router.push(transaction);
```
Comment (vinh0604) [inline pat-link-sample/src/pages/index.tsx:38]: We should use `window.localtion.href = ...` here to trigger redirect.

File: script/deploy-template.json [hunk 6]
```diff
@@ -56,6 +56,12 @@
       "ports": {
         "3004": "HTTP"
       }
+    },
+    "merchant-pat-link": {
+      "image": "$MERCHANT_PAT_LINK_IMAGE_NAME",
```
Comment (vinh0604) [inline script/deploy-template.json:61]: We also need to update the `deploy.sh` to add this new image to the deployment step.

**Review by vinh0604 — CHANGES_REQUESTED** [review]

File: pat-link-sample/src/pages/index.tsx (hunk 5 above)
Comment (loiopenfabric) [reply pat-link-sample/src/pages/index.tsx:38]: It does not work

Comment (vinh0604) [reply pat-link-sample/src/pages/index.tsx:38]: Please try 
```ts
const transaction = await (
  await fetch("/api/link", ...)
).json();

window.location.href = transaction.url;
```

**Review by vinh0604 — APPROVED** [review]
LGTM

**Pull request merged**